snuba/optimize.py

```diff
@@ -292,18 +341,58 @@ def optimize_partitions(
 
         query = (query_template % args).strip()
         logger.info(f"Optimizing partition: {partition}")
-        start = time.time()
 
         # Update tracker before running clickhouse.execute since its possible
         # that the connection can get disconnected while the server is still
         # executing the optimization. In that case we don't want to
         # run optimization on the same partition twice.
         if tracker:
-            tracker.update_completed_partitions(partition)
+            tracker.update_scheduled_partitions(partition)
 
+        # if theres a merge in progress for this partition, wait for it to finish
+        while True:
+            busy_merging, estimated_sleep_time = is_busy_merging(
+                clickhouse, database, table
+            )
+            if not busy_merging:
+                break
+            else:
+                time.sleep(OPTIMIZE_BASE_SLEEP_TIME + estimated_sleep_time)
+
+        start = time.time()
         clickhouse.execute(query, retryable=False)
         metrics.timing(
             "optimized_part",
             time.time() - start,
             tags=_get_metrics_tags(table, clickhouse_host),
         )
+
+
+def is_busy_merging(
+    clickhouse: ClickhousePool, database: str, table: str
+) -> Tuple[bool, float]:
+    """
+    Returns true and the estimated sleep time if clickhouse is busy with merges in progress
+    for the table. Clickhouse is considered busy if
+        1. there are more than OPTIMIZE_MERGE_MAX_LONG_CONCURRENT_JOBS merges in progress with an elapsed time greater than OPTIMIZE_MERGE_MIN_ELAPSED_CUTTOFF_TIME
```

Comment:
Mind cutting this line shorter. It is really hard to read. According to PEP-8 docstrings should be 80 characters wide so they remain more readable on code reviews.